src/ClimaAtmos/Parameterizations/TurbulenceConvection/src/StateVecs/StateVecFuncs.jl

```diff
@@ -0,0 +1,322 @@
+
+Formulaically, a total covariance between variables ``ϕ`` and ``ψ`` is computed from
+
+``⟨ϕ^*ψ^*⟩ = Σ_i a_i \\overline{ϕ_i'ψ_i'} + Σ_i Σ_j a_i a_j \\overline{ϕ}_i (\\overline{ψ}_i - \\overline{ψ}_j)``
+
+Where variable ``\\overline{ϕ}_i`` represents ``ϕ`` decomposed across multiple
+sub-domains, which are weighted by area fractions ``a_i``.
+"""
+function total_covariance!(dst::StateVec, src::StateVec, cv::StateVec, weights::StateVec,
+                           dst_idxs, cv_idxs, weight_idx, grid::Grid, decompose_ϕ_ψ::Function)
+  for k in over_elems(grid), (dst_idx, cv_idx) in zip(dst_idxs, cv_idxs)
+    _ϕ, _ψ = decompose_ϕ_ψ(cv_idx)
+    dst[dst_idx, k] = 0
+    for i in over_sub_domains(weights)
+      dst[dst_idx, k] += weights[weight_idx, k, i]*cv[cv_idx, k, i]
+      for j in over_sub_domains(src)
+        dst[dst_idx, k] += weights[weight_idx, k, i]*
+                           weights[weight_idx, k, j]*
+                           src[_ϕ, k, i]*(src[_ψ, k, i] - src[_ψ, k, j])
+      end
+    end
+  end
+end
+
+"""
+    extrap!(sv::StateVec, name::Symbol, grid::Grid, i_sd=1)
+
+Extrapolate variable `name` to the first ghost point.
+"""
+function extrap!(sv::StateVec, name::Symbol, grid::Grid, i_sd=1)
+  k = 1+grid.n_ghost
+  sv[name, k-1, i_sd] = 2*sv[name, k, i_sd] - sv[name, k+1, i_sd]
+  k = grid.n_elem - grid.n_ghost
+  sv[name, k+1, i_sd] = 2*sv[name, k, i_sd] - sv[name, k-1, i_sd]
+end
+
+"""
+    assign_ghost!(sv::StateVec, name::Symbol, val, grid::Grid, i_sd=1)
+
+Assign value `val` to variable `name` for all ghost points.
+"""
+function assign_ghost!(sv::StateVec, name::Symbol, val, grid::Grid, i_sd=1)
+  for k in over_elems_ghost(grid)
+    sv[name, k, i_sd] = val
+  end
+end
+
+"""
+    integrate_ode!(sv::StateVec,
+                   name::Symbol,
+                   grid::Grid,
+                   func::Function,
+                   y_0::AbstractFloat,
+                   args::NamedTuple,
+                   i_sd::Int = 1)
+
+Integrates the ODE:
+  `dy/dz = func(z, args)`
+  `y = y_0 + int_{z=0}^{z} func(z, args) dz`
+"""
+function integrate_ode!(sv::StateVec,
+                        name::Symbol,
+                        grid::Grid,
+                        func::Function,
+                        y_0::AbstractFloat,
+                        args::NamedTuple,
+                        i_sd::Int = 1)
+  k = first_elem_above_surface(grid)
+  sv[name, k-1, i_sd] = 0
+  for k in over_elems_real(grid)
+    f = func(get_z(grid, k), args)
+    sv[name, k, i_sd] = sv[name, k-1, i_sd] + grid.dz * f
+  end
+  for k in over_elems_real(grid)
+    sv[name, k, i_sd] += y_0
+  end
+  assign_ghost!(sv, name, 0.0, grid, i_sd)
+end
+
+abstract type ExportType end
+"""
+    UseVTK
+
+A singleton used to indicate to use a .vtk
+extension when exporting a `StateVec`.
+"""
+struct UseVTK <: ExportType end
+
+"""
+    UseDat
+
+A singleton used to indicate to use a .dat
+extension when exporting a `StateVec`.
```

Comment:
I don't think `.dat` is a standard file format, maybe explain what it outputs.

Reply:
Added a bit more info, let me know if you want more. I'd be fine changing this to `.csv` if that's easier.